java/com/google/re2j/Unicode.java

```diff
@@ -227,6 +229,99 @@ static int simpleFold(int r) {
     return toUpper(r);
   }
 
+  static int[] optimizedFoldOrbit(int r) {
+    if (r >= ORBIT_MIN_VALUE) {
+      return FOLD_MAP.get(r);
+    } else return null;
+  }
+
+  static int optimizedFoldNonOrbit(int r) {
+    int l = toLower(r);
+    if (l != r) {
+      return l;
+    }
+    return toUpper(r);
+  }
+  
+  private static final FoldMap FOLD_MAP = new FoldMap(UnicodeTables.CASE_ORBIT.length * 4);
+  private static final int ORBIT_MIN_VALUE = UnicodeTables.CASE_ORBIT[0][0];
+  
+  static {
+    for(int i = 0; i < UnicodeTables.CASE_ORBIT.length; i++) {
+      int r0 = UnicodeTables.CASE_ORBIT[i][0];
+      int[] folds = new int[0];
+      int r1 = r0;
+      while((r1 = simpleFold(r1)) != r0) {
+        folds = Arrays.copyOf(folds, folds.length + 1);
```

Comment:
Isn't an orbit necessarily of length at least 3? I realize they are small and few, but it seems perverse to copy the entire array each time we add an element.  Use a temporary large enough for the longest orbit (MAX_DISTANCE?) and copy it once before each call to FOLD_MAP.put.